sktime/decomposition/vampnet.py

```diff
@@ -0,0 +1,662 @@
+from ..util import module_available
+
+if not module_available("torch"):
+    raise RuntimeError("Tried importing VampNets; this only works with a PyTorch installation!")
+del module_available
+
+from typing import Optional, Union, List, Callable, Tuple
+
+import numpy as np
+
+from ..base import Transformer, Model, Estimator
+from ..data.util import TimeSeriesDataset
+
+import torch
+import torch.nn as nn
+from torch.utils.data import Dataset, DataLoader
+
+
+def sym_inverse(mat, epsilon: float = 1e-6, ret_sqrt=False, mode='regularize'):
+    """ Utility function that returns the inverse of a matrix, with the
+    option to return the square root of the inverse matrix.
+
+    Parameters
+    ----------
+    mat: numpy array with shape [m,m]
+        Matrix to be inverted.
+    epsilon : float
+        Cutoff for eigenvalues.
+    ret_sqrt: bool, optional, default = False
+        if True, the square root of the inverse matrix is returned instead
+    mode: str, default='trunc'
+        Whether to truncate eigenvalues if they are too small or to regularize them by taking the absolute value
+        and adding a small positive constant. :code:`trunc` leads to truncation, :code:`regularize` leads to epsilon
+        being added to the eigenvalues after taking the absolute value
+
+    Returns
+    -------
+    x_inv: numpy array with shape [m,m]
+        inverse of the original matrix
+    """
+    assert mode in sym_inverse.valid_modes, f"Invalid mode {mode}, supported are {sym_inverse.valid_modes}"
+
+    if mode == 'trunc':
+        # Calculate eigvalues and eigvectors
+        eigval_all, eigvec_all_t = torch.symeig(mat, eigenvectors=True)
+        # Filter out Eigenvalues below threshold and corresponding Eigenvectors
+        mask = eigval_all > epsilon
+
+        eigval = eigval_all[mask]
+        eigvec = eigvec_all_t.transpose(0, 1)[mask]
+        eigvec_t = eigvec.transpose(0, 1)
+    elif mode == 'regularize':
+        # Calculate eigvalues and eigvectors
+        identity = torch.eye(mat.shape[0], dtype=mat.dtype, device=mat.device)
+        eigval, eigvec_t = torch.symeig(mat + epsilon*identity, eigenvectors=True)
+        eigval = torch.abs(eigval)
+        eigvec = eigvec_t.transpose(0, 1)
+    elif mode == 'clamp':
+        eigval, eigvec_t = torch.symeig(mat, eigenvectors=True)
+        eigval = torch.clamp_min(eigval, min=epsilon)
+        eigvec = eigvec_t.transpose(0, 1)
+    else:
+        raise RuntimeError("invalid mode!")
```

Comment:
is this error descriptive enough that if it gets raised people will know the mode that's invalid was for this fxn?

Reply:
this error is caught by the assertion further up in the method's implementation and should never really raise. Added it so that my IDE does not complain :)